PR: Add qtsass compilation step

goanpeca opened the pull request:
Fixes #137
Fixes #80

- [x] Add tests
- [x] Update package dependencies
- [x] Release qtsaas pypi package
- [x] Release qtsaas conda-forge package

goanpeca commented:
@dpizetta I still need to make the pypi release of qtsass and make the conda-forge package, but this is the general idea. Thoughts, suggestions?

ccordoba12 commented:
This looks really nice, thanks @goanpeca! @dpizetta, as you know, we need this to get QDarkstyle colors programmatically to make some adjustments in Spyder.

@goanpeca, a couple of comments:

1. This probably needs to be done against the `dev` branch, which has some work already for QDarkstyle 3.0
2. I think there's no need to maintain auto-generated files in the repo. So `_variables.scss` and `style.qss` should be removed and instead generate them by extending the `build` command of `setup.py`, so they are included in the source and wheel tarballs.
3. Tests can cover that the generated files are built correctly by `qtsaas`.

dpizetta commented:
Thank you @goanpeca. About the branch, it is ok to add in the v2.x. I need to move all fixes to the v3 also. Other points agreed with @ccordoba12.

Any ideas to keep scripts running when testing changes locally without reinstallation? May they need an update.

PS: I've published a release with some corrections. Some update is needed.

Thanks for your help!

goanpeca commented:
Regarding:

>This probably needs to be done against the dev branch, which has some work already for QDarkstyle 3.0

Missed that!

>I think there's no need to maintain auto-generated files in the repo. So _variables.scss and style.qss should be removed and instead generate them by extending the build command of setup.py, so they are included in the source and wheel tarballs.

I am half half here, I know that the autogenerated files should probably not be kept in version control, but it makes evaluating changes and understanding what is going on a bit easier.

>Any ideas to keep scripts running when testing changes locally without reinstallation? May they need an update.

For the `qss`, sure but for the resto of the generated py files I am not sure.

Importing QDarkStyle could rerender everything, so when testing I guess this would suffice (if only using the QSS ?)

goanpeca commented:
>Any ideas to keep scripts running when testing changes locally without reinstallation? May they need an update.

Was this a general question or specific to this PR ?

If working on dev mode (`dev` in version we could instantiate an `Observer`?)

dpizetta commented:
> Missed that!, on further inspection, no changes seem to collide :-)

For now, I'll keep in version 2. There are more things to be done there.

> I am half half here, I know that the autogenerated files should probably not be kept in version control, but it makes evaluating changes and understanding what is going on a bit easier.

Same as you. No big problems according to the size, as we kept the compile ui files too. See the next one.

> For the qss, sure but for the resto of the generated py files I am not sure. Importing QDarkStyle could rerender everything, so when testing I guess this would suffice (if only using the QSS ?)

If they are generated (.qss) when importing, everything is ok. I need to try here :) Other ones are generated (ui>py, qss>py). There is no need to keep these generate files, but avoid using things that are not so easy to install in all versions (pyside-rcc, pyqtuic, etc).  I've been doing a commit with the readable changes and then a commit to update compiled files. If better ideas show up, we can change.

> Was this a general question or specific to this PR ?

Specific to this. I mean, how I got generated the (old) .qss files as the scripts run the compilation for each binding? For example, here https://github.com/ColinDuquesnoy/QDarkStyleSheet/blob/ac0929d43239c1dfe632264e14072312f0749122/script/process_qrc.py#L73.
May this is a dumb question as I'm not familiar with SSAS before. :)

goanpeca commented on qdarkstyle/utils.py:
```diff
@@ -0,0 +1,93 @@
+def create_qss():
+    """Create variables files and run qtsass compilation."""
+    _create_scss_variables()
+    _create_qss()
```
@dpizetta this method runs things. This is what I “intended to run on import

Thoughts?

dpizetta replied:
Every time you import you will call this function to update the qss file. When the qss file would be generated to then be compiled into .py files. (I need to familiarize with scss dynamic)

goanpeca replied:
how expensive is the qss to .py conversion?

dpizetta replied:
> how expensive is the qss to .py conversion? and is this something we can rely will work on any user machine with any of the bindings?

It is not so expansive, but it is not so easy to use the tool to convert qss > py. We were providing the compiled files to simplify the use of qdarkstyle. Each binding could generate to itself, but most of then must be installed with dev-tools to integrate the compilers. 

> I mean we don't have to do this right now if we want to keep QDarkStyle more "static" but moving forward we could use this to provide users a simple way to style apps beyond DarkStyles by changing the variables (and maybe extra ones we can define).

Agreed totally. My idea is still providing the compiled files to the default qdarkstyle values. But generating the qss using your solution with scss. Then we can work on the changeable qdarkstyle colors on the "develper/user" machines. So I think it is a good idea, for now, stay with qss and compiled files in the repo. I was reading about loading ui, qss and other files directly in the python bindings, but I'm not sure about how trustful they are. I had many problems in the past with this, including open .png files (not converted to binaries). For now we can stay with nsert .scss (repo) to generate .qss (repo) and provide compiled forms .py (repo).

Opinions?

goanpeca replied:
I will add the compilation step to the scripts (so they are compiled before doing the other conversion) and a small test and then wait for the package (qtsass) to be released so the test can work, and it can be added as a dependency of QDarkStyle. Will do that later today ;-).

After that I want to automate the SVG - PNG generation from the palette colors as well.

dpizetta replied:
The scripts in the script folder do all needed until now, just run them inside the script folder, just add the qtass to them. (I don't know if you say exactly the same thing auhau). I have all the tools installed here for qt4/qt5, I can easily generate them. 

Go ahead! Thanks a lot!

goanpeca commented:
@ccordoba12 although https://github.com/sass/libsass-python supports python 2.7, the conda forge recipe skips for windows and py27 https://github.com/conda-forge/libsass-feedstock.

What do you think we should do then ? make libsass 

https://github.com/conda-forge/staged-recipes/pull/8201/files

goanpeca commented:
Ok so I will release the conda forge package soon. For winpy27 no packages so we can still use the bundled one when making releases (that is why I prefer to leave the generated files here as well)

ccordoba12 commented:
> What do you think we should do?

We should try to compile it with support for Python 2, it shouldn't be that hard.

goanpeca commented:
Libsass does not work with vs2008 it seems

dpizetta commented:
Also, this can work on C++ files? Many projects use QdarkStyle in C++. At least we can generate the qss file that is loaded on C++ directly.

ColinDuquesnoy commented:
@goanpeca @dpizetta 

Having sass is a good idea but we absolutely need to keep generated qss and png files (#142) in the repository for users of Qt that are not using python. Many users just drop the content of qdarkstyle directory into their own project and build the qrc file as part of their projects. Most of them do not have python installed and do not want an additional dependency. But I guess some users will want to benefit from this PR and customize the style to fit their own UI/UX guidelines and branding colors. So I think it would be a good idea to clearly documents the tools and the procedure needed to customize the stylesheet after a fresh clone of the repo. The documentation should be written for users who are not used to the python ecossytem.

Regarding the ability to customize the stylesheet at runtime when using the python package: I suggest a different approach:

1. Embed sass and svg files in the resources
2. Write sass and svg files from qrc to a temp directory
3. Generate qss and png files in the temp directory
4. Replace qrc:// url in the stylesheet by the temp directory
5. Load the stylsheet from the temporary dir instead of the the qrc file.

I strongly suggest this to be a separate API with optional dependencies to qtsass,... If a user wants to load the official stylesheet the above process should not be executed and no dependency to qtsass is needed.

goanpeca commented:
>Also, this can work on C++ files? Many projects use QdarkStyle in C++. At least we can generate the qss file that is loaded on C++ directly.

@dpizetta that is why the idea is to leave the (generated) files with the repo

goanpeca commented:
>Having sass is a good idea but we absolutely need to keep generated qss and png files (#142) in the repository for users of Qt that are not using python. Many users just drop the content of qdarkstyle directory into their own project and build the qrc file as part of their projects. Most of them do not have python installed and do not want an additional dependency. But I guess some users will want to benefit from this PR and customize the style to fit their own UI/UX guidelines and branding colors. So I think it would be a good idea to clearly documents the tools and the procedure needed to customize the stylesheet after a fresh clone of the repo. The documentation should be written for users who are not used to the python ecossytem.

Hi @ColinDuquesnoy agreed :+1:

>Regarding the ability to customize the stylesheet at runtime when using the python package: I suggest a different approach:

>Embed sass and svg files in the resources
>Write sass and svg files from qrc to a temp directory
>Generate qss and png files in the temp directory
>Replace qrc:// url in the stylesheet by the temp directory
>Load the stylsheet from the temporary dir instead of the the qrc file.

Yes, this is something similar to what I had in mind ;-), where I would change the temo dir, for something the user can select (preferably using the home directory as a default, not a temp dir)

>I strongly suggest this to be a separate API with optional dependencies to qtsass,... If a user wants to load the official stylesheet the above process should not be executed and no dependency to qtsass is needed.

Agreed, there should be a split. Thanks for the input

Regarding the dependencies, I would still add qtsass as a normal dependency, even if only needed for advanced usage. Thoughts there?

ColinDuquesnoy commented:
> Regarding the dependencies, I would still add qtsass as a normal dependency, even if only needed for advanced usage. Thoughts there?

If I understand correctly, for now qtsass is only needed for developers so I would add it to the dev dependencies. Once it is used in the end-user package for the advanced feature we talked about, yes I would add it to the runtime dependencies.

goanpeca commented:
@dpizetta I incorporated the latest fixes in master to keep the generated qss in sync.

- I added the `PATH_RESOURCES` variable.
- I added a check for qtsass so that if present it will run the compilation otherwise do nothing (instead of raising an exception).
- I created another issue to track the creation of custom stylesheets at runtime (which is out of the scope of this PR)
- I added qtsass as a dev dependency

This one should be ready now :-)

dpizetta commented:
Perfect! Tks!
